universal-application-tool-0.0.1/app/repository/ProgramRepository.java

```diff
@@ -159,40 +164,93 @@ public Program createOrUpdateDraft(Program existingProgram) {
+            .in(
+                "lifecycle_stage",
+                ImmutableList.of(LifecycleStage.ACTIVE, LifecycleStage.OBSOLETE));
+
+    if (searchNameFragment.isPresent() && !searchNameFragment.get().isBlank()) {
+      String search = searchNameFragment.get();
+
+      if (search.matches("^\\d+$")) {
+        query = query.or().eq("id", search).eq("applicant_id", search).endOr();
+      } else {
+        search = search.toLowerCase(java.util.Locale.ROOT);
+
+        query =
+            query
+                .or()
+                .eq("submitter_email", search)
+                .raw(
+                    getApplicationObjectPath("applicant", "name", "first_name") + " ILIKE ?",
+                    "%" + search + "%")
+                .raw(
+                    getApplicationObjectPath("applicant", "name", "last_name") + " ILIKE ?",
+                    "%" + search + "%")
+                .endOr();
+      }
+    }
+
+    PagedList<Application> pagedQuery =
+        query
+            .setFirstRow((paginationSpec.getCurrentPage() - 1) * paginationSpec.getPageSize())
+            .setMaxRows(paginationSpec.getPageSize())
+            .findPagedList();
+
+    pagedQuery.loadCount();
+
+    return new PaginationResult<Application>(
+        paginationSpec,
+        pagedQuery.getTotalPageCount(),
+        pagedQuery.getList().stream().collect(ImmutableList.toImmutableList()));
+  }
+
+  private Query<Program> allProgramVersionsQuery(long programId) {
     return database
         .find(Program.class)
         .select("id")
         .where()
-        .eq(
-            "name",
-            database.find(Program.class).setId(programId).select("name").findSingleAttribute())
-        .findSingleAttributeList()
-        .stream()
-        .map((obj) -> (Long) obj)
-        .collect(ImmutableList.toImmutableList());
+        .in("name", database.find(Program.class).select("name").where().eq("id", programId).query())
+        .query();
   }
 
-  public ImmutableList<Application> getApplicationsForVersions(ImmutableList<Long> versionIds) {
-    return database
-        .find(Application.class)
-        .fetch("program")
-        .where()
-        .in("program_id", versionIds)
-        .or()
-        .eq("lifecycle_stage", LifecycleStage.ACTIVE)
-        .eq("lifecycle_stage", LifecycleStage.OBSOLETE)
-        .endOr()
-        .findList()
-        .stream()
-        .collect(ImmutableList.toImmutableList());
+  private String getApplicationObjectPath(String... pathComponents) {
+    StringBuilder path = new StringBuilder();
+
+    path.append("(");
+
+    // While the column type is JSONB, CiviForm writes the JSON object into the database as a
+    // string.
+    // This requires queries that interact with the JSON column to instruct postgres to parse the
+    // contents
+    // into JSON, which we do here with (object #>> '{}')::jsonb
+    path.append("(object #>> '{}')::jsonb");
+
+    for (int i = 0; i < pathComponents.length - 1; i++) {
```

Comment:
optional. capture the last index in a var to reduce future change errors between the two uses.

Reply:
done